habitat-lab/habitat/articulated_agent_controllers/humanoid_base_controller.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+
+# Copyright (c) Meta Platforms, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+
+import magnum as mn
+import numpy as np
+
+
+class Pose:
+    def __init__(self, joints_quat, root_transform):
+        """
+        Contains a single humanoid pose
+            :param joints_quat: list or array of num_joints * 4 elements, with the rotation quaternions
+            :param root_transform: Matrix4 with the root trasnform.
+        """
+        self.joints = list(joints_quat)
+        self.root_transform = root_transform
+
+
+class Motion:
+    """
+    Contains a sequential motion, corresponding to a sequence of poses
+        :param joints_quat_array: num_poses x num_joints x 4 array, containing the join orientations
+        :param transform_array: num_poses x 4 x 4 array, containing the root transform
+        :param displacement: on each pose, how much forward displacement was there?
+            Used to measure how many poses we should advance to move a cerain amount
+        :param fps: the FPS at which the motion was recorded
+    """
+
+    def __init__(self, joints_quat_array, transform_array, displacement, fps):
+        num_poses = joints_quat_array.shape[0]
+        self.num_poses = num_poses
+        poses = []
+        for index in range(num_poses):
+            pose = Pose(
+                joints_quat_array[index].reshape(-1),
+                mn.Matrix4(transform_array[index]),
+            )
+            poses.append(pose)
+
+        self.poses = poses
+        self.fps = fps
+        self.displacement = displacement
+
+
+class HumanoidBaseController:
+    """
+    Generic class to replay SMPL-X motions
+        :param draw_fps: the FPS at which we should be advancing the pose.
```

Comment:
maybe not "draw_fps" since that implies rendering which is not controlled here.

Reply:
Resolved